src/libsyntax/parse/parser.rs

```diff
@@ -7769,26 +7767,22 @@ impl<'a> Parser<'a> {
             let ident = self.parse_ident()?;
             if self.check(&token::OpenDelim(token::Brace)) {
                 // Parse a struct variant.
-                all_nullary = false;
                 let (fields, recovered) = self.parse_record_struct_body()?;
                 struct_def = VariantData::Struct(fields, recovered);
             } else if self.check(&token::OpenDelim(token::Paren)) {
-                all_nullary = false;
                 struct_def = VariantData::Tuple(
                     self.parse_tuple_struct_body()?,
                     ast::DUMMY_NODE_ID,
                 );
-            } else if self.eat(&token::Eq) {
+            } else {
+                struct_def = VariantData::Unit(ast::DUMMY_NODE_ID);
+            }
```

Comment:
This `if-else` chain can be put in a `let struct_def = ...;` now, to simplify it.